ouroboros-network/src/Ouroboros/Network/Mux/Interface.hs

```diff
@@ -0,0 +1,164 @@
+    :: forall ps (st :: ps) m a.
+       Codec ps CBOR.DeserialiseFailure m ByteString
+    -> Peer ps AsClient st m a
+    -> MuxPeer m
+
+  -- |
+  -- A pipelined peer together with a codec.
+  --
+  OnlyPipelinedClient
+    :: forall ps (st :: ps) m a.
+       Natural
+    -> Codec ps CBOR.DeserialiseFailure m ByteString
+    -> PeerPipelined ps AsClient st m a
+    -> MuxPeer m
+
+  -- |
+  -- Server peer with a codec
+  OnlyServer
+    :: forall ps (st :: ps) m a.
+       Codec ps CBOR.DeserialiseFailure m ByteString
+    -> Peer ps AsServer st m a
+    -> MuxPeer m
+
+  -- |
+  -- Client and server peers with the corresponding codec.
+  --
+  ClientAndServer
+    :: forall ps (st :: ps) m a.
+       Codec ps CBOR.DeserialiseFailure m ByteString
+    -> Peer ps AsClient st m a
+    -> Peer ps AsServer st m a
+    -> MuxPeer m
+
+  -- |
+  -- Pipelined client and a server with the correspnding codec.
+  --
+  PipelinedClientAndServer
+    :: forall ps (st :: ps) m a.
+       Natural
+    -> Codec ps CBOR.DeserialiseFailure m ByteString
+    -> PeerPipelined ps AsClient st m a
+    -> Peer ps AsServer st m a
+    -> MuxPeer m
+
+
+-- |
+-- Public network interface for 'ouroboros-network'.
+--
+data NetworkInterface ptcl addr m = NetworkInterface {
+      -- |
+      -- Address of the node to run.  The node will bind to this address, and
+      -- listen for incoming connections.  Some bearers do not have a notion of
+      -- address.
+      --
+      nodeAddress :: addr, 
+
+      -- |
+      -- Versioned set of peers to run; it is the responsibilty of the caller
+      -- to provide the correct associtation between `ptcl` and a protocol
+      -- (which is existentially quantified using @'PeerSpecification'@).
+      --
+      protocols   :: [(Version, ptcl -> MuxPeer m)]
+    }
+
+-- | Low level network interface.  It can be intiatiated using a socket, pair
+-- of pipes or a pair queues.
+--
+data NetworkNode addr m = NetworkNode {
+      -- |
+      -- The way to connect ot other peers.  On startup the network interface
+      -- will run this to connect with a given list of peer.  But it can also
+      -- be used at a later stage to connect to a new peer.
+      --
+      -- This function will run client side of mux version negotation and then
+      -- start a the list protocols given by @'NetworkInterface'@.
+      --
+      connectTo :: addr -> m (),
+
+      -- |
+      -- This will cancel the thread that is listening for new connections and
+      -- close the underlaying bearer.
+      killNode  :: m ()
+    }
+
+-- |
+-- Transform a @'MuxPeer'@ into @'ProtocolDescription'@ used by the
+-- multiplexing layer.
+--
+miniProtocolDescription
+  :: forall m ptcl.
+     ( MonadAsync m
+     , MonadCatch m
+     , MonadThrow m
+     )
+  => MuxPeer m
+  -> MiniProtocolDescription ptcl m
+miniProtocolDescription (OnlyClient codec peer) =
+  MiniProtocolDescription {
+      mpdInitiator = Just (\chan -> void (runPeer codec chan peer)),
```

Comment:
Something we're going to need here eventually is the ability to run some code that brackets the `runPeer`. For example for block fetch each node needs to register itself with the thread that does the block fetch decisions. 
```
bracketFetchClient :: FetchClientRegistry peer header m
                   -> peer
                   -> (FetchClientStateVars header m -> m a)
                   -> m a
```
Oh, and this highlights another thing, some of the protocol handlers need to know the id of the peer they are talking to. The block fetch client does, and so does the chain sync client. (They both do things like updating the state in a `Map` entry for their `peer`.)

We will have to consider how to do both of these things.

If we resolve that, that might also resolve your problem with the mux not returning values, but without having to change the protocols themselves (like the req/resp) to add terminal effects.

Reply:
So your `bracketFetchClient` is basically a custom `runPeer` function (It must call it inside); we could make it an argument of the `MuxPeer`, making it in a CPS style
```
data MuxPeer a where
   OnlyClient :: Codec -> Peer -> (Channel -> Codec -> Peer -> m a) -> MuxPeer
   ...
```
This would only partially solve the problem of the returned value by the mux layer: `a` would be common for all the protocols, but since it's only for tests maybe it's justified; Otherwise we'd need a type level function that maps `ptcl` to a right tuple (we'd need to make a distinction between `OnlyClient`, `OnlyServer` using a GADT as well and this mapping would need to take this in account, since the result will depend on wheather one is running only client or server or both). This seems possible, but maybe this would be over complex.